<h1 style="color: #001a79;">Weekly Exercises - 01 Statistics</h1>

<hr style="border-top: 1px solid #001a79;" />

<h3 style="color: #001a79;">Exercise 1.1</h3>

<hr style="border-top: 1px solid #001a79;" />

__Task:__
<br>

- The above gives about a 1.4% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.

The following code is adapted from:  https://stackoverflow.com/a/4941932

Python version 3.8 includes a new method which [geeksforgeeks](https://www.geeksforgeeks.org/python-math-comb-method/) explains can be used to return the number of ways to choose k items from n items. This method works without repetition and without order.
<br><br>
*Syntax: math.comb(n, k)*

In [1]:
# Math module.
import math

In [2]:
# Number of ways of selecting 4 cups from 8.
math.comb(8, 4)

70

In [3]:
# Number of ways of selecting 5 cups form 10.
math.comb(10, 5)

252

In [4]:
# Probability of randomly selecting correct 5 cups is:
1.0 / math.comb(10, 5)

0.003968253968253968

So, if the experiment was done with 10 cups total, five with milk in first and five with tea in first, then the chance of  randomly selecting correctly is much less than 1%.

We could design the experiment to have 9 cups in total, with 4 with milk in first and 5 with tea in first - or vice versa. 

In [5]:
# Number of ways of selecting 4 cups from 9. 
math.comb(9, 4)

126

In [6]:
# Number of ways of selecting 5 cups from 9:
math.comb(9, 5)

126

Both have the same number of ways of selecting cups. This is because the two groups of choices remain the same, just in a different order. Therefore, the outcome will have the same number of ways of selecting cups regardless of the which group selects first. 
<br>

In [7]:
# The probability is then:
1.0 / math.comb(9, 4)

0.007936507936507936

This example is also less than 1%.
<br>
<br>
<h4 style="color: #001a79;">However,  the question is what is the minimun number of cups required to ensure the probability of randomly selecting the correct one is less than or equal to 1%. </h4>
<br>
The answer would depend on the sample size. Take the below cells, The probability of selecting 5 cups from 10 correctly is less than 1%. The probability of selecting 3 cups from 9 is about  1.1% and selecting 4 cups of 8 correctly is about 1.4%. <br>

My theory is that the higher the sample size, leads to more ways of selecting cups, which in turn leads to a lower probability of selecting correctly. Therefore, if we wanted to ensure a <=1% chance of choosing correctly I would go with a sample of 10. There are more possibilities which in turn creates a lower probability of choosing correctly. 
<br>

In [8]:
# Number of ways of selecting 5 cups from 10.
a = math.comb(10, 5)

# Probability of randomly selecting correctly.
b = 1.0 / a

a, b

(252, 0.003968253968253968)

In [9]:
# Number of ways of selecting 3 cups from 9.
c = math.comb(9, 3)

# Probability of randomly selecting correctly.
d = 1 / c

c, d

(84, 0.011904761904761904)

In [10]:
# Number of ways of selecting 4 cups from 8.
e = math.comb(8, 4)

# Probability of randomly selecting correctly.
f = 1 / e

e, f

(70, 0.014285714285714285)

<hr style="border-top: 1px solid #001a79;" />
<br>
<h3 style="color: #001a79;">Exercise 1.2</h3>

<hr style="border-top: 1px solid #001a79;" />

__Bonus Task:__
<br>

- How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?

In [11]:
# Number of ways of selecting. 
a = math.comb(10, (5-1))

# Probability of randomly selecting correctly after getting one wrong.
b = 1.0 / a

a, b

(210, 0.004761904761904762)

My answer is to continue to use a sample of 10 while deducting the "wrong answer". As we can see there are still 210 ways of selecting 4 cups from 10 after getting one wrong which is still producing a less than 1% probibility of choosing the rest correctly. 
<br>

<hr style="border-top: 1px solid #001a79;" />
<br>
<h3 style="color: #001a79;">Exercise 1.3</h3>
<hr style="border-top: 1px solid #001a79;" />

__Task:__
<br>

- Use <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html" style="color: #ff791e">scipy's version of Fisher's exact test</a> to simulate the Lady Tasting Tea problem.
<br>

In [12]:
# Statistical methods. 
import scipy.stats as ss

In [13]:
# We can see the syntax of this function takes in a two sided table.
ss.fisher_exact

<function scipy.stats.stats.fisher_exact(table, alternative='two-sided')>

The desired outcome is:
<br>

*0.014285714285714285*

In [14]:
# Testing on a sample table.
ss.fisher_exact([[4, 0], [0, 4]])

(inf, 0.028571428571428536)

In [15]:
# Desired outcome double is almost the same as fisher's exact test. 
0.014285714285714285 * 2

0.02857142857142857

In [16]:
# Testing again with added parameter. 
table = [[4, 0], [0, 4]]
oddsr, p = ss.fisher_exact(table, alternative='greater')
p

0.014285714285714268

This piece of code has given the exact answer by using the parameter *alternative='greater'*. This is an optional parameter that gives odds for the probability being greater than one. 

In other words, the fisher exact method is used to determine if there is a significant association between two variables. The variables are either independent or they're not. What this code means is there is less than 1% chance of the probabilty being greater than 1. 

I found two interesting blog posts which explains the fisher_exact method which I found helpful to understand the statistical aspect of the experiment. They can be found here: 
<br>

* https://towardsdatascience.com/fishers-exact-test-from-scratch-with-python-2b907f29e593
<br>

* https://www.statology.org/fishers-exact-test/#:~:text=Fisher%27s%20Exact%20Test%20is%20used,table%20is%20less%20than%205
<br>

<hr style="border-top: 1px solid #001a79;" />
<br>
<h3 style="color: #001a79;">Exercise 1.4</h3>
<hr style="border-top: 1px solid #001a79;" />

__Task:__
<br>

- Take the code from the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html" style="color: #ff791e">Examples section of the scipy stats documentation for independent samples t-tests</a>, add it to your own notebook and add explain how it works using MarkDown cells and code comments. Improve it in any way you think it could be improved.

In [17]:
# For numerical arrays.
import numpy as np

# Using rng as reccommended to generate random default float as state. 
rng = np.random.default_rng()

# For statistical functions.
from scipy import stats

In [18]:
# Test with a sample with identical means:
# Note the random state is set as reccommended. 

# Norm.rvs to calculate the noraml distribution
rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)

# ttest_ind 
stats.ttest_ind(rvs1, rvs2)

Ttest_indResult(statistic=0.008987947092030064, pvalue=0.9928305486109621)

What this means is that scipy stats norm is used to analyze the normal distribution. Within this module there are various methods. One of these is rvs. Rvs returns random varieties. In other words, this stats.norm.rvs method has randomly generated values which were instances of normal distribution. This is calculated using the mean and standard deviation which in this example is 5 and 10 respectively. The size is referring to the size of the output array ie. output array would contain 500 values. 
<br>

Explaination found here: https://www.delftstack.com/api/scipy/scipy-scipy.stats.norm-method/
<br>

Moreover, the stats.ttest_ind is used. The offical documentationa around scipy stats is quite comprehensive in my opinion, however, after reviewing it I found a good comment regarding t-tests:
<br>

*The t-test quantifies the difference between the arithmetic means of the two samples*
<br>

In a nutshell, a t-test calculates the means of two independent samples. In this example, we already determined the mean and standard deviation and recieved random varieties from the stats.norm.rvs method. Now, we are passing these two variables into the t-test. 
<br>

In [19]:
stats.ttest_ind(rvs1, rvs2, equal_var=False)

Ttest_indResult(statistic=0.008987947092030064, pvalue=0.9928305487386876)

In the next example there is an option parameter present. Basically, if True it will perform the t-test asssuming equal population variances. This is the default option. Otherwise, if false, it performs a Welch’s t-test. This is a t-test but with unequal variances. In other words, there are unequal sample sizes. This may be useful if for example we were testing a sample of 9 - the two independent samples have different sizes as one may be 5 and the other 4. Thus, these samples would not have normal distribution. This can be seen when comparing the results of this cell to the one previous where the t-test used the default paramater of True suggesting normal distribution of samples. 
<br>

Wikipedia for the Welch’s t-test found here: https://en.wikipedia.org/wiki/Welch%27s_t-test

In [20]:
# Another example of a t-test.
stats.ttest_ind(rvs1, rvs2, permutations=10000, random_state=rng)

Ttest_indResult(statistic=0.008987947092030064, pvalue=0.993)

This is another example of the t-test except the size now reffered to as permutations is a lot larger than the previous examples. The documentation explains that more permutations returns more accurate results. This may be because a larger sample can give more results which can be analyzed on a larger scale giving a clear picture of results. 
<br>

In [21]:
# Passing in values instead of using numpy to generate. 
a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
b = (1.1, 2.9, 4.2)

# Doing a t-test without getting normal distribution. 
stats.ttest_ind(a, b, trim=.2)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)

<h4 style="color: #001a79;">Conclusion</h4>

<br>

A t-test is a statistical test. It is used to analyze the means of two samples or the mean and standard deviation of two samples. It is a test for the null hypothesis that two independent samples have the same average values. 
<br>

<hr style="border-top: 1px solid #001a79;" />

<h3 style="color: #001a79;">End</h3>